### PR TITLE
fix(docker): kill the in-container process tree on terminal teardown

### LINKED ISSUE #84967: [Bug]: Docker terminal timeout leaves in-container process trees running
## Bug Description

With `terminal.backend: docker`, a terminal timeout or interrupt kills the local `docker exec` CLI process but does not reliably terminate the command tree that `docker exec` started inside the container. The in-container shell, test runner, and multiprocessing workers can survive as orphans. Repeated timeouts can exhaust the container's pids cgroup, after which even trivial terminal calls such as `date` or `pwd` hang.

This was reproduced on the latest release, v0.20.0 (`v2026.8.3`). The affected code path is still present on current `origin/main` at `67168a391f26211eba13adfe4ecee74bd365a25b`.

This is distinct from #68915: that issue concerned shell `&` backgrounding and inherited stdout. Here, an ordinary foreground command exceeds the terminal timeout on the Docker backend.

## Steps to Reproduce

1. Configure a persistent Docker terminal backend.
2. Run a foreground command that starts descendants and exceeds the terminal timeout. A minimal synthetic example is:

   ```bash
   python -c 'import subprocess,time; subprocess.Popen(["sh","-c","sleep 600"]); time.sleep(600)'
   ```

3. Let the Hermes terminal tool return exit 124 due to its outer timeout.
4. From the Docker host, inspect the target container with `docker top` or `docker exec ... ps`.
5. Repeat the timed-out call several times, or use a multiprocessing workload such as pytest + PyTorch DataLoader workers.

## Expected Behavior

On timeout, interrupt, cancellation, or agent shutdown, Hermes should terminate and reap the complete in-container process tree for that terminal call (bounded `SIGTERM`, then `SIGKILL`), without affecting unrelated commands in the persistent container.

## Actual Behavior

The local `docker exec` client exits, while processes started inside the container can continue running and become reparented/orphaned.

Sanitized evidence from the reproduced incident:

```text
8 surviving pytest command roots were found after the original full-suite call timed out
pids.max: 256
pids.peak: 257
pids.events: max 1344
```

At that point, new terminal calls such as `date` and `pwd` timed out after 10-60 seconds. After terminating only the orphaned pytest/DataLoader trees, fresh exec calls completed in 2-7 ms again. The Docker engine itself remained healthy throughout.

## Affected Component

- Tools (terminal)
- Docker backend

## Debug Report

Not attached because the live profile's config/log bundle contains unrelated private workspace and credential metadata. The sanitized process/cgroup evidence and source-level root cause are included here.

## Environment

- macOS 26.5.1 (arm64)
- OrbStack Docker-compatible runtime
- Docker Engine 29.4.0
- Python 3.11.15
- Hermes Agent v0.20.0 (`v2026.8.3`)
- Persistent Docker terminal container

## Root Cause Analysis

On current `main`:

- `tools/environments/docker.py::_run_bash()` returns a local `subprocess.Popen` for `docker exec ... bash -c <command>`.
- `tools/environments/base.py::_wait_for_process()` handles timeout/interrupt by calling `self._kill_process(proc)`.
- The base `_kill_process()` implementation only calls `proc.kill()`.
- `DockerEnvironment` does not override this to identify and terminate the corresponding process tree inside the container.

Therefore, Hermes kills the host-side Docker CLI process, not necessarily the command descendants in the container.

## Proposed Fix

Give each Docker terminal invocation a unique in-container lifecycle boundary and clean it up explicitly:

1. Start each exec command in a uniquely identifiable process group/session (or equivalent per-exec cgroup).
2. On timeout, interrupt, cancellation, nonzero launcher failure, or shutdown, signal that in-container group with bounded `SIGTERM` followed by `SIGKILL`.
3. Reap/verify that no descendants for the invocation remain.
4. Keep cleanup scoped so unrelated concurrent commands in the shared persistent container are not killed.
5. Add regression tests with a launcher that spawns descendants, including a multiprocessing worker tree, and assert no survivors after timeout.

Related: #68915 fixed a different local/background-shell variant of process leakage; this Docker remote-exec path remains uncovered.

### PR DESCRIPTION
**Pull request by deepspace28**
Fixes #84967.

## The bug

`BaseEnvironment._kill_process()` is `proc.kill()` and nothing else — and its own docstring says *"Subclasses may override for process-group kill."* `DockerEnvironment` never did, so on timeout the kill lands on the host-side `docker exec` client while the command's descendants keep running inside the container. They get reparented and hold pids-cgroup slots; enough timed-out calls and `pids.max` is exhausted, after which even `date` or `pwd` hangs.

## Why tag the environment instead of killing a process group

`docker exec` never reports the in-container PID back to the client, so the host has no PGID to signal — the `setsid` route would need the launcher to write its PGID somewhere and the host to read it back, adding a round trip and a failure mode.

Each exec is instead stamped with a unique `HERMES_EXEC_ID` and the tree is recovered by scanning `/proc/*/environ`. That is what makes it correct in the case that actually leaks:

- the environment is **inherited by every descendant**, so multiprocessing workers and nested trees are caught, not just direct children;
- it **survives reparenting** — the leaked processes have already been adopted away from the exec root, so anything walking parent PIDs has lost them, but they still carry the marker;
- it is **inherently scoped** — only processes bearing that exact id are signalled, so unrelated commands sharing a persistent container are never touched (point 4 of the issue).

`_run_bash` already injects `-e` args for the profile-scoped passthrough, so this rides an existing pattern.

## The change

`DockerEnvironment._kill_process()` keeps the host-side `proc.kill()`, then runs a bounded `SIGTERM` → wait → `SIGKILL` teardown inside the container. All three base-class kill paths — timeout, interrupt, exception — funnel through this one override, so it covers the full set the issue asks for. Cleanup is best-effort and never raises: a teardown failure must not mask the condition that triggered the kill.

The script is pure bash deliberately. `_run_bash` already requires bash, whereas `pgrep` is absent from many minimal images, `ps` flags differ under busybox, `grep -z` is GNU-only, and busybox `sleep` may reject fractional seconds.

One defensive note: handles that don't accept attributes (a bare `object()` has no `__dict__`) fall back to the previous host-only kill rather than raising on the spawn path.

## Verification

Run in a Linux container, since the container side is always Linux:

- **113 passed** across `test_docker_environment.py`, `test_docker_session_isolation.py`, `test_docker_cgroup_limits.py`, `test_base_environment.py`, `test_docker_orphan_reaper_integration.py` and the new file — no regressions.
- The new file passes **3 consecutive runs** with no flakes.
- With the fix reverted, **5 of the 8 new tests fail** (the 3 that still pass are the no-target/failure-isolation guards, which pass vacuously against the base implementation).

Two of the new tests are the real evidence and run only on Linux:

- `test_tree_kill_script_reaps_reparented_descendants` — the launcher backgrounds a sleep and exits, so the survivor is reparented before teardown runs; asserts it is still found and killed.
- `test_tree_kill_script_leaves_unrelated_processes_alone` — asserts a differently-tagged process beside it survives.

They signal through the script itself rather than `os.kill`, so they stay inside the conftest live-system guard instead of reaching for the documented bypass marker.

## Not covered

I haven't run the issue's full multiprocessing repro (pytest + PyTorch DataLoader workers) against a real persistent backend under load — the synthetic reparented-orphan case is what I could verify directly. Happy to add cgroup-based per-exec scoping instead if you'd prefer that over environment tagging; it's a design call.

### CODE REVIEW

**Comment by deepspace28** [comment]
Ping — `main` has moved ~1274 commits since this opened and the branch is still conflict-free.

This is the one of my open PRs I'd most like a look at, because the failure it fixes is silent: on terminal teardown only the direct child is signalled, so anything it spawned survives inside the container and keeps holding resources. Nothing surfaces an error — the processes just accumulate.

Happy to rebase, or to trim the tests down if the diff size is what's making it awkward to pick up.

**Comment by Enough1122** [comment]
> AI code review — automated review for reference, author can ignore or act on any point.

**fix(docker): kill the in-container process tree on terminal teardown**

1. `tools/environments/docker.py` — `_TREE_KILL_SCRIPT` scans `/proc/[0-9]*` for the marker exactly **once** at the top. Any descendant that forks *after* that scan (during the SIGTERM grace window) inherits `HERMES_EXEC_ID` but never enters `$pids`, so it survives the final SIGKILL pass and keeps holding a pids-cgroup slot — the exact leak class this PR fixes. Consider re-scanning the marker inside the grace loop (or looping the whole scan until nothing tagged remains), and add a test where a tagged process forks a child mid-grace.

2. Every `_kill_process` now costs an extra `docker exec` round-trip plus a container-wide `/proc/*/environ` read. On a busy shared persistent container the full scan is O(all container processes) on a teardown path; bounded and best-effort, but worth a note that a per-exec teardown session (or skipping the scan when the handle shows the exec client never really started) could reduce it.

3. `test_tree_kill_script_reaps_reparented_descendants` spawns a childless `sleep 600`; the forked-during-grace case above (the more adversarial shape) is untested — the re-scan fix would be easy to pin with the existing `_marked_pids` helper.

4. Minor: `_kill_container_tree(grace=...)` is Python-side only; the script's own loop sleeps in whole seconds, so a fractional grace silently rounds down — fine at the default 2s, just be aware if it's ever made configurable.

Nice property worth keeping: the teardown exec itself is untagged, so the tree-kill script never kills itself.
